### DISCUSSION TITLE
User is created but password couldn't be updated

### DISCUSSION THREAD

**Discussion by akulavasu** [Sunbird-Ed/Community#678] [Sunbird ED configuration discussion] [answered]
Hi @santhosh-tg  @gandham-santhosh ,

On hitting the POST request to User create Admin (/api/user/v1/create), we get the following error

Request body:
```
{
    "request": {
        "firstName": "SunbirdOrgAdmin",
        "lastName": "",
        "email": "SunbirdOrgAdmin@yopmail.com",
        "userName": "{{auth_username}}",
        "password": "{{auth_password}}",
        "emailVerified":true,
        "dob":"2001",
        "channel":"sunbird-xi",
        "roles": ["ORG_ADMIN", "CONTENT_CREATOR", "CONTENT_REVIEWER"]
    }
}
```

Response
```
{
    "id": "api.user.create",
    "ver": "v1",
    "ts": "2024-01-19 05:52:00:559+0000",
    "params": {
        "resmsgid": "0638543ee0b1643680fcf032b82e64d7",
        "msgid": "0638543ee0b1643680fcf032b82e64d7",
        "err": null,
        "status": "SUCCESS",
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "userId": "cce0eb7b-e2c1-4ac2-9024-36214ae35c25",
        "response": "SUCCESS",
        "err_msg": "User is created but password couldn't be updated.",
        "errors": []
    }
}
```

As mentioned in https://github.com/orgs/Sunbird-Ed/discussions/306 , we checked the following

1. Cassandra provider is available in user federation(Keycloak sunbird realm)
2. Channel is available in cassandra DB. 
3. Player service --> Realm is set to sunbird in configmap **but there is no key for SSO URL.**
4. Set the below variables to false in player configmap
sunbird_p1_reCaptcha_enabled: false
sunbird_p2_reCaptcha_enabled: false
sunbird_p3_reCaptcha_enabled: false

6. Verified all these variables are updated in private repo - 
core_vault_sunbird_sso_publickey
adminutil_refresh_token_public_key_kid
adminutil_refresh_token_secret_key

7. Assigned manage-users role to lms client

Please let us know what configuration is missing.

**Comment by gandham-santhosh** [comment]
@akulavasu were you not able to see this user in the system after it is created? Did you check if there are any errors seen on learner service? @AmiableAnil

**Reply by akulavasu** [reply]
@gandham-santhosh  - Yeah, the user is not getting reflected in keycloak. 

This is the error in learner service

`{"timestamp":"2024-01-22T04:56:50.146Z","msg":"{\"eid\":\"ERROR\",\"ets\":1705899410146,\"ver\":\"3.0\",\"mid\":\"21c1936661d72daaa961d915c6ed254c\",\"actor\":{\"id\":\"db9b82cf-ff3d-4d8c-a69a-6dbab7a1b47c\",\"type\":\"Consumer\"},\"context\":{\"channel\":\"01397213214902681610\",\"pdata\":{\"id\":\"dev.sunbird.learning.service\",\"pid\":\"learner-service\",\"ver\":\"5.3.0\"},\"env\":\"User\",\"cdata\":[{\"id\":\"21c1936661d72daaa961d915c6ed254c\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"err\":\"UOS_0069\",\"stacktrace\":\"updatePassword: Exception occurred:  org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine.invoke(ApacheHttpClient4Engine.java:289)org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:454)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:105)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)com.sun.proxy.$Proxy35.grantToken(Unknown Source)org.keycloak.admin.client.token.TokenManager.grantToken(TokenManager.java:89)org.keycloak.admin.client.token.TokenManager.getAccessToken(TokenManager.java:69)org.keycloak.admin.client.token.TokenManager.getAccessTokenString(TokenManager.java:64)org.keycloak.admin.client.resource.BearerAuthFilter.filter(BearerAuthFilter.java:52)org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:431)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:105)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)com.sun.proxy.$Proxy50.resetPassword(Unknown Source)org.sunbird.sso.impl.KeyCloakServiceImpl.updatePassword(KeyCloakServiceImpl.java:75)org.sunbird.util.user.UserUtil.updatePassword(UserUtil.java:298)org.sunbird.service.user.impl.SSOUserServiceImpl.createUserAndPassword(SSOUserServiceImpl.java:99)org.sunbird.actor.user.SSOUserCreateActor.processSSOUser(SSOUserCreateActor.java:117)org.sunbird.actor.user.SSOUserCreateActor.createSSOUser(SSOUserCreateActor.java:89)org.sunbird.actor.user.SSOUserCreateActor.onReceive(SSOUserCreateActor.java:61)org.sunbird.actor.core.BaseActor.onReceive(BaseActor.java:21)akka.actor.UntypedAbstractActor$$anonfun$receive$1.applyOrElse(AbstractActor.scala:291)akka.actor.Actor$class.aroundReceive(Actor.scala:517)akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:180)akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)akka.actor.ActorCell.invoke(ActorCell.scala:557)akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)akka.dispatch.Mailbox.run(Mailbox.scala:225)akka.dispatch.Mailbox.exec(Mailbox.scala:235)akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\",\"errtype\":\"api_access\",\"requestid\":\"21c1936661d72daaa961d915c6ed254c\"}}","lname":"TelemetryEventLogger","tname":"application-akka.actor.most-used-one-dispatcher-94","level":"INFO","HOSTNAME":"learner-7465cfbff7-ml2sx","application.home":"/home/sunbird/learner/learning-service-1.0-SNAPSHOT"}`

**Reply by akulavasu** [reply]
@gandham-santhosh - Any update here would be helpful in completion of validating Lern API.

**Reply by AmiableAnil** [reply]
@akulavasu Ensured all the prerequisites are met with respect to secrets as discussed in the thread https://github.com/orgs/Sunbird-Ed/discussions/306. Any suggestions on what could be the cause for this error.

**Reply by akulavasu** [reply]
@AmiableAnil ,  Verified all the variables are provided in secrets and checked the same in learner configmap.
The users are getting created in cassandra user_organisation table. 
How can we verify if user federation with cassandra is configured correctly.

**Reply by santhosh-tg** [reply]
@akulavasu  In Keycloak -> Cassandra-storage-provider settings can you confirm if the Cassandra host is point to correct IP ?

**Reply by akulavasu** [reply]
@santhosh-tg - Yeah, it's pointing to the correct IP.

**Reply by akulavasu** [reply]
Below tables in keycloak postgres DB are empty. Is that expected.
![image](https://github.com/Sunbird-Ed/Community/assets/34706681/234b9498-9a05-4a02-916b-224ea00a03a2)

**Reply by santhosh-tg** [reply]
@akulavasu Yes, that is expected. Can you run the user create API again and check both learner and keycloak logs for any errors?

**Reply by akulavasu** [reply]
@santhosh-tg , There is no new log created in keycloak when running user create API. 
In learner service, we get the below attached error.
[learner-error-log.txt](https://github.com/Sunbird-Ed/Community/files/14033833/learner-error.txt)

**Reply by santhosh-tg** [reply]
Hi @AmiableAnil @Hari-stackroute Request you to take a look at the logs and share your findings.

**Reply by AmiableAnil** [reply]
@akulavasu Can we have a call today?

cc: @santhosh-tg @Hari-stackroute

**Reply by akulavasu** [reply]
@AmiableAnil - Yes, we can

**Reply by AmiableAnil** [reply]
@akulavasu we can connect here at 4:00PM

Microsoft Teams meeting
Join on your computer, mobile app or room device
[Click here to join the meeting](https://teams.microsoft.com/l/meetup-join/19%3ameeting_ZGE4N2Y2MzUtOTgzMy00NDMwLWJmNDUtNTFmMzlhZDAxNjU2%40thread.v2/0?context=%7b%22Tid%22%3a%22d2f80462-bf28-4055-8ef3-085d8e4b2ed0%22%2c%22Oid%22%3a%222d4e8def-ebcf-474e-a0ac-63e56966a02a%22%7d)
Meeting ID: 446 030 820 79
Passcode: Gf8dKv
[Download Teams](https://www.microsoft.com/en-us/microsoft-teams/download-app) | [Join on the web](https://www.microsoft.com/microsoft-teams/join-a-meeting)
[Learn More](https://aka.ms/JoinTeamsMeeting) | [Meeting options](https://teams.microsoft.com/meetingOptions/?organizerId=2d4e8def-ebcf-474e-a0ac-63e56966a02a&tenantId=d2f80462-bf28-4055-8ef3-085d8e4b2ed0&threadId=19_meeting_ZGE4N2Y2MzUtOTgzMy00NDMwLWJmNDUtNTFmMzlhZDAxNjU2@thread.v2&messageId=0&language=en-IN)

**Reply by akulavasu** [reply]
Thanks for the call @AmiableAnil . As discussed, let's connect again to resolve the same.

Issues observed so far,
1. Temporary password not getting updated in keycloak.
2. After manually providing credential in keycloak and logging in portal shows Access denied and below error is logged in player service.

```
{"channel":"sunbird","env":"dev.dev.portal","level":"info","message":{"msg":"KONG_TOKEN :: request using existing token","originalUrl":"/resources?auth_callback=1&state=dd08127e-70af-47c1-8fad-174610f829f8&session_state=22ca50be-d5da-4daa-a34b-cba2906c200d&code=6871a73c-e71a-40f4-9f49-eb5d318571f7.22ca50be-d5da-4daa-a34b-cba2906c200d.8c12290d-d62f-48ce-913b-c93bf995ca59","route":"/resources","sessionId":"3ceXAZMnUU2lx5XjiceGDiXBueoeMQ6y"},"pdata":{"id":"dev.dev.portal","pid":"sunbird-portal-backend","ver":"6.0.0"},"timestamp":"2024-01-29 11:35:17"}
Error from postauth > getGrantFromCode {
  err: Error: unable to verify the first certificate
      at TLSSocket.onConnectSecure (node:_tls_wrap:1539:34)
      at TLSSocket.emit (node:events:513:28)
      at TLSSocket._finishInit (node:_tls_wrap:953:8)
      at TLSWrap.ssl.onhandshakedone (node:_tls_wrap:734:12) {
    code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'
  },
  queryObj: {
    auth_callback: '1',
    state: 'dd08127e-70af-47c1-8fad-174610f829f8',
    session_state: '22ca50be-d5da-4daa-a34b-cba2906c200d',
    code: '6871a73c-e71a-40f4-9f49-eb5d318571f7.22ca50be-d5da-4daa-a34b-cba2906c200d.8c12290d-d62f-48ce-913b-c93bf995ca59'
  }
}
Could not obtain grant code: Error: unable to verify the first certificate
```

**Reply by akulavasu** [reply, marked as answer]
@santhosh-tg  @gandham-santhosh @AmiableAnil , the second issue mentioned above (Unable to verify the first certificate) is resolved now after updating the certificate chain crt file along with the site certificate crt in the core_vault_proxy_site_crt variable under core/secrets.yml in private repo.

Once this was fixed, the keycloak password update also started working as expected.

Thanks
